### DISCUSSION TITLE
A constructor is a static method

### DISCUSSION THREAD

**Discussion by wrexbe** [dotnet/csharplang#7415] [Language Ideas]
The general idea, is to be able to use constructors like any other static method that returns a new instance of a class, possibly with some tuple magic added in.

Before:
```
int port = 9092;
var handlers = Helpers
    .GetValidListeningIPv4Addresses()
    .Select(ipAddress => new TcpListener(ipAddress, port))
    .Select(listener => new TcpListenerHandler(listener))
    .ToList();
```

After
```
int port = 9092;
var handlers = Helpers
    .GetValidListeningIPv4Addresses() // Method that gets IPs I want to bind to.
    .Select(address => (address, port))
    .Select(new TcpListener) // Calls constructor TcpListener(IPAddress localaddr, int port)
    .Select(new TcpListenerHandler) // Calls user defined constructor TcpListenerHandler(TcpListener listener)
    .ToList();
```

I also want a way to treat constructors like any other static method

```
Func<object> = new object;

being an easier, and maybe more performant of this

Func<object> = MakeObject;

public static object MakeObject()
{
    return new object();
}
```

The exact syntax isn't the point, it is the ability to use a constructor like other static methods.

**Comment by jnm2** [comment]
Constructors are implemented in the .NET runtime as void-returning instance methods. This is how `: this(...)` and `: base(...)` call chaining is implemented.

**Reply by TahirAhmadov** [reply]
They are implemented as instance methods, but the question, what's preventing the language from compiling the examples in the OP? The title is perhaps worded incorrectly, should say "Constructors should be usable like static methods", but the text of the OP does not claim they are _actually_ static methods.

**Reply by saint4eva** [reply]
What you need is static constructor.

**Comment by HaloFour** [comment]
The C# language could support syntax like this, by emitting a bridge method that calls the constructor.  This is not unlike constructor references in Java, which can be used in lambdas:

```java
Callable<Object> action = Object::new;

Object obj = action.call();
```

Of course, it's also pretty trivial to invoke the constructor within the body of a lambda.

**Reply by TahirAhmadov** [reply]
Yes this would be essentially syntactic sugar. OP's concern with performance is unfounded, though, because both a manually written static method and the compiler generated bridge static method would be most likely optimized away anyway. Having said that, I agree this OP is doable.

**Reply by wrexbe** [reply]
The performance concern was mostly if you do the lambda way , not wrapping with a static method

**Comment by iam3yal** [comment]
I think the motivation here is to allow to pass constructors as method group to delegates but then it's fairly simple to introduce a static `Create` method to do that but requires slightly more work and for existing code that don't have these static factory methods obviously it wouldn't work unless you create extensions, wrappers or whatever but if you do have control over the code then something like this:
```
int port = 9092;
var handlers = Helpers
    .GetValidListeningIPv4Addresses() // Method that gets IPs I want to bind to.
    .Select(address => (address, port))
    .Select(TcpListener.Create) // Calls constructor TcpListener(IPAddress localaddr, int port)
    .Select(TcpListenerHandler.Create) // Calls user defined constructor TcpListenerHandler(TcpListener listener)
    .ToList();
```

**Comment by HaloFour** [comment]
This proposal also implies tuple splatting, to translate the tuple argument into an argument list:

See:
https://github.com/dotnet/csharplang/discussions/424
https://github.com/dotnet/csharplang/discussions/1654

**Comment by HaloFour** [comment]
Also, existing discussion about treating a constructor as a method group:

#674

**Comment by wrexbe** [comment]
The alternatives have down sides.

- They won't adapt if you want to use a different overload, you would need to write more of them, or change the existing ones
- They are boilerplate (that get worse the more arguments, and overloads an object has)
- If you want to write a static method to keep performance, you need to go to another place to change the method.
- Writing a lambda to wrap a single method call feels bad, when you don't have to do that for any other method.

The tuple splatting thing is neat, but treating a constructor as an implicit Func<TParam1,TParam2,TOutput>, would be nice.

**Reply by iam3yal** [reply]
@wrexbe Well you can either write a SG to automate it or wait for the future if it ever gets implemented.
